README.md
## Posa Bottom Bar 🔥

- made by Hmida71
- in <a href='https://pub.dev/packages/posa_bottom_bar'>https://pub.dev/packages/posa_bottom_bar</a>



### ANIMATION EXAMPLE

| Animation Icons | Animation elevation and opacity |
|---------|----------|
|![FanBottomNavyBar Gif](https://user-images.githubusercontent.com/69757558/200888205-fbdd698c-39ae-4b04-a0d1-4560d746f83f.gif) | ![Fix Gif](https://user-images.githubusercontent.com/69757558/200888217-37199c10-6186-4ab5-b254-c44bec68daec.gif) |


## Customization

### PosaBottomBar (REQUIRED ITEMS)

- `child` - add here your page or navigator pages.
- `icons` - list of items (IconData), required more than one item and less than six.
- `onChanged` - required to listen when an item is tapped it provides the selected item's index.

### PosaBottomBar (OPTIONAL ITEMS)

- `onLongPress` - required to listen when an item is longPressed it provides the selected item's index.
- `iconSize` - the item icon's size.
- `dotSize` - bottom(bottom of icon) items size.
- `backgroundColor` - the navigation bar's background color.
- `elevation` - if make it 0 elevation will be removed.
- `shadowColor` - color of elevation bottom bar.
- `borderRadius` - radius of bottom bar.
- `height` - changes the Navigation Bar's height.
- `margin` - changes the Navigation Bar's margin.
- `alignment` - use this property to change the vertical alignment of the bottom bar. It is mostly used when you have two (center) or (centerBottom) or (centerTop) the bar.
- `isElevationAnimation` - if make it true we get somme animation of elevation.
- `elevationAnimationOpacity` - if change if to > 0.3 the color of elevation becomes too dark.
- `elevationAnimationDuration` - param to customize the item change's animation duration
- `isAnimatedIcon` - if make it true we get somme animation icons position

## Getting Started

Add the dependency in `pubspec.yaml`:

```yaml
dependencies:
  posa_bottom_bar: ^latest_version
```

## Basic Usage

Adding the widget

```dart
@override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text("POSA BOTTOM BAR"),
        centerTitle: true,
      ),
      body: PosaBottomBar(
        // CHILD IS REQUIRED
        child: pages
            .elementAt(currentIndex), // ADD HERE YOUR PAGE OR PAGES NAVIGATOR
        // ICONS IS REQUIRED
        icons: const [
          Ionicons.home_outline,
          Ionicons.settings_outline,
          Ionicons.bookmark_outline,
          Ionicons.person_outline,
        ],
        // ONCHANGE IS REQUIRED
        onChanged: (i) {
          setState(() {
            debugPrint("TAP $i");
            currentIndex = i;
          });
        },
        // ON LONG PRESS FNC
        onLongPress: (i) {
          setState(() {
            debugPrint("LONG PRESS $i");
          });
        },
        alignment: Alignment.bottomCenter,
        // ------ height and margin for bottom bar ----- //
        height: 0.16,
        margin: 0.05,
        // ------  ----- //

        // ------ Elevation & Color for bottom bar ----- //
        elevation: 10,
        shadowColor: Colors.black,
        // ------  ----- //

        borderRadius: 35, // RADUIS (FOR BOTTOM BAR)
        backgroundColor: Colors.blueAccent, // BACKGROUND COLORS (FOR BOTTOM BAR)
        dotColor: Colors.white, // COLORS OF DOT
        iconColor: Colors.white, // COLORS OF ICON
        dotSize: 5.0, iconSize: 24.0, // SIZE OF DOT AND ICONS

        // ------ Animation elevation and opacity ----- //
        isElevationAnimation: false,
        elevationAnimationOpacity: 0.3,
        elevationAnimationDuration: const Duration(milliseconds: 1500),
        // ------  ----- //

        // ------ Animation Icons ------ //
        isAnimatedIcon: false,
        // ------  ------ //
      ),
    );
  }
```


CHANGELOG.md
## 0.0.1

- Bottom Bar Navigation Widget With Cool Animation | Flutter

## 0.0.2

- Update readme.md (file) | Flutter

## 0.0.3

- Fix some bug | Flutter

example/lib/page_test.dart
import 'dart:io';

import 'package:first_packeges/main.dart';
import 'package:flutter/material.dart';
import 'package:ionicons/ionicons.dart';
import 'package:url_launcher/url_launcher.dart';

class PageTest extends StatefulWidget {
  const PageTest({Key? key}) : super(key: key);

  @override
  State<PageTest> createState() => _PageTestState();
}

class _PageTestState extends State<PageTest> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SafeArea(
        child: Center(
            child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            Container(
                margin: const EdgeInsets.only(left: 30, bottom: 20),
                child: Text(
                  "Page ${currentIndex + 1} test yoo !",
                  style: const TextStyle(fontSize: 20),
                )),
            InkWell(
              splashColor: Colors.purple[200],
              onTap: () async {
                Uri urldiscord = Uri.parse("https://github.com/Hmida71");

                if (Platform.isAndroid) {
                  await launchUrl(urldiscord,
                      mode: LaunchMode.externalApplication);
                }
              },
              child: ListTile(
                selectedTileColor: Colors.purple[200],
                selectedColor: Colors.purple[200],
                isThreeLine: false,
                leading: const CircleAvatar(
                    radius: 25,
                    child: Icon(
                      Ionicons.logo_github,
                      size: 28,
                    )),
                title: const Text("Check Github : Hmida71"),
                subtitle: const Text("Web : https://hmida71.netlify.app"),
              ),
            )
          ],
        )),
      ),
    );
  }
}


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility in the flutter_test package. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:first_packeges/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(const MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


lib/posa_bottom_bar.dart
library posa_bottom_bar;

import 'package:flutter/material.dart';
import 'package:animator/animator.dart';

class PosaBottomBar extends StatefulWidget {
  final Widget? child;
  final Function(dynamic i) onChanged;
  final List<IconData> icons;
  final Function(dynamic i)? onLongPress;
  final Color? backgroundColor;
  final Color? iconColor;
  final Color? dotColor;
  final Color? shadowColor;
  final double? elevation;
  final double? borderRadius;
  final double? iconSize;
  final double? dotSize;
  final double? margin;
  final double? height;
  final AlignmentGeometry? alignment;
  final bool? isAnimatedIcon;
  final bool? isElevationAnimation;
  final Duration? elevationAnimationDuration;
  final double? elevationAnimationOpacity;

  const PosaBottomBar({
    Key? key,
    required this.child,
    required this.onChanged,
    required this.icons,
    this.onLongPress,
    this.backgroundColor,
    this.iconColor,
    this.dotColor,
    this.shadowColor = Colors.black,
    this.elevation,
    this.borderRadius = 35,
    this.iconSize = 24.0,
    this.dotSize = 5.0,
    this.margin = .05,
    this.height = .160,
    this.alignment = Alignment.bottomCenter,
    this.isAnimatedIcon = false,
    this.isElevationAnimation = false,
    this.elevationAnimationDuration = const Duration(milliseconds: 1500),
    this.elevationAnimationOpacity = 0.3,
  }) : super(key: key);

  @override
  State<PosaBottomBar> createState() => _PosaBottomBarState();
}

class _PosaBottomBarState extends State<PosaBottomBar>
    with TickerProviderStateMixin {
  int currentIndex = 0;

  late AnimationController _animationController;
  late Animation _animation;

  @override
  void initState() {
    if (widget.icons.length > 6) {
      widget.icons.removeRange(6, widget.icons.length);
    }
    if (widget.isElevationAnimation!) {
      _animationController = AnimationController(
        vsync: this,
        duration: widget.elevationAnimationDuration!,
      );
      _animationController.repeat(reverse: true);
      _animation = Tween(begin: 1.0, end: 13.0).animate(_animationController)
        ..addListener(
          () {
            setState(() {});
          },
        );
    }
    super.initState();
  }

  @override
  void dispose() {
    _animationController.dispose();
    _animation.isDismissed;
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    double displayWidth = MediaQuery.of(context).size.width;
    return Scaffold(
        body: Stack(
      alignment: widget.alignment!,
      children: [
        widget.child!,
        normalBar(displayWidth),
      ],
    ));
  }

  Widget normalBar(displayWidth) {
    return Container(
      margin: EdgeInsets.all(displayWidth * widget.margin!),
      child: Material(
        elevation: widget.elevation ?? 0,
        shadowColor: widget.shadowColor,
        borderRadius: BorderRadius.all(Radius.circular(widget.borderRadius!)),
        child: Container(
          height: displayWidth * widget.height!,
          alignment: Alignment.center,
          decoration: BoxDecoration(
              color: widget.backgroundColor ?? const Color(0xff8a2be1),
              borderRadius:
                  BorderRadius.all(Radius.circular(widget.borderRadius!)),
              boxShadow: widget.isElevationAnimation!
                  ? [
                      BoxShadow(
                        color: widget.shadowColor!
                            .withOpacity(widget.elevationAnimationOpacity!),
                        blurRadius: _animation.value + _animation.value,
                        spreadRadius: _animation.value,
                        offset: Offset(_animation.value, _animation.value),
                      ),
                    ]
                  : []),
          child: Row(
            children: [
              const SizedBox(width: 4),
              ...List.generate(widget.icons.length, (i) {
                return Expanded(
                  child: GestureDetector(
                    onTap: () => setState(() {
                      currentIndex = i;
                      widget.onChanged(i);
                    }),
                    onLongPress: () => setState(() {
                      currentIndex = i;
                      widget.onLongPress!(i);
                    }),
                    child: Container(
                      color: Colors.transparent,
                      margin: const EdgeInsets.symmetric(
                          horizontal: 4, vertical: 2),
                      height: displayWidth * .155,
                      child: Column(
                        mainAxisAlignment: MainAxisAlignment.center,
                        children: [
                          widget.isAnimatedIcon!
                              ? currentIndex == i
                                  ? Animator<double>(
                                      duration:
                                          const Duration(milliseconds: 2000),
                                      cycles: 0,
                                      curve: Curves.easeInOut,
                                      tween:
                                          Tween<double>(begin: 0.0, end: 5.0),
                                      builder:
                                          (context, animatorState, child) =>
                                              Column(
                                        children: [
                                          SizedBox(
                                            height: animatorState.value * 2.5,
                                          ),
                                          Container(
                                            decoration: const BoxDecoration(
                                                shape: BoxShape.circle,
                                                color: Colors.transparent),
                                            child: Icon(
                                              widget.icons[i],
                                              color: widget.iconColor ??
                                                  Colors.white,
                                              size: widget.iconSize,
                                            ),
                                          ),
                                          const SizedBox(height: 6),
                                          currentIndex == i
                                              ? Container(
                                                  width: widget.dotSize,
                                                  height: widget.dotSize,
                                                  decoration: BoxDecoration(
                                                    shape: BoxShape.circle,
                                                    color: widget.dotColor ??
                                                        Colors.white,
                                                  ),
                                                )
                                              : Container(),
                                          const SizedBox(
                                            height: 2,
                                          ),
                                        ],
                                      ),
                                    )
                                  : Icon(
                                      widget.icons[i],
                                      color: widget.iconColor ?? Colors.white,
                                      size: widget.iconSize,
                                    )
                              : Column(
                                  children: [
                                    Icon(
                                      widget.icons[i],
                                      color: widget.iconColor ?? Colors.white,
                                      size: widget.iconSize,
                                    ),
                                    currentIndex == i
                                        ? Container(
                                            width: widget.dotSize,
                                            height: widget.dotSize,
                                            margin:
                                                const EdgeInsets.only(top: 6),
                                            decoration: BoxDecoration(
                                              shape: BoxShape.circle,
                                              color: widget.dotColor ??
                                                  Colors.white,
                                            ),
                                          )
                                        : Container(),
                                  ],
                                ),
                        ],
                      ),
                    ),
                  ),
                );
              }),
              const SizedBox(width: 4),
            ],
          ),
        ),
      ),
    );
  }
}


example/linux/my_application.cc
#include "my_application.h"

#include <flutter_linux/flutter_linux.h>
#ifdef GDK_WINDOWING_X11
#include <gdk/gdkx.h>
#endif

#include "flutter/generated_plugin_registrant.h"

struct _MyApplication {
  GtkApplication parent_instance;
  char** dart_entrypoint_arguments;
};

G_DEFINE_TYPE(MyApplication, my_application, GTK_TYPE_APPLICATION)

// Implements GApplication::activate.
static void my_application_activate(GApplication* application) {
  MyApplication* self = MY_APPLICATION(application);
  GtkWindow* window =
      GTK_WINDOW(gtk_application_window_new(GTK_APPLICATION(application)));

  // Use a header bar when running in GNOME as this is the common style used
  // by applications and is the setup most users will be using (e.g. Ubuntu
  // desktop).
  // If running on X and not using GNOME then just use a traditional title bar
  // in case the window manager does more exotic layout, e.g. tiling.
  // If running on Wayland assume the header bar will work (may need changing
  // if future cases occur).
  gboolean use_header_bar = TRUE;
#ifdef GDK_WINDOWING_X11
  GdkScreen* screen = gtk_window_get_screen(window);
  if (GDK_IS_X11_SCREEN(screen)) {
    const gchar* wm_name = gdk_x11_screen_get_window_manager_name(screen);
    if (g_strcmp0(wm_name, "GNOME Shell") != 0) {
      use_header_bar = FALSE;
    }
  }
#endif
  if (use_header_bar) {
    GtkHeaderBar* header_bar = GTK_HEADER_BAR(gtk_header_bar_new());
    gtk_widget_show(GTK_WIDGET(header_bar));
    gtk_header_bar_set_title(header_bar, "first_packeges");
    gtk_header_bar_set_show_close_button(header_bar, TRUE);
    gtk_window_set_titlebar(window, GTK_WIDGET(header_bar));
  } else {
    gtk_window_set_title(window, "first_packeges");
  }

  gtk_window_set_default_size(window, 1280, 720);
  gtk_widget_show(GTK_WIDGET(window));

  g_autoptr(FlDartProject) project = fl_dart_project_new();
  fl_dart_project_set_dart_entrypoint_arguments(project, self->dart_entrypoint_arguments);

  FlView* view = fl_view_new(project);
  gtk_widget_show(GTK_WIDGET(view));
  gtk_container_add(GTK_CONTAINER(window), GTK_WIDGET(view));

  fl_register_plugins(FL_PLUGIN_REGISTRY(view));

  gtk_widget_grab_focus(GTK_WIDGET(view));
}

// Implements GApplication::local_command_line.
static gboolean my_application_local_command_line(GApplication* application, gchar*** arguments, int* exit_status) {
  MyApplication* self = MY_APPLICATION(application);
  // Strip out the first argument as it is the binary name.
  self->dart_entrypoint_arguments = g_strdupv(*arguments + 1);

  g_autoptr(GError) error = nullptr;
  if (!g_application_register(application, nullptr, &error)) {
     g_warning("Failed to register: %s", error->message);
     *exit_status = 1;
     return TRUE;
  }

  g_application_activate(application);
  *exit_status = 0;

  return TRUE;
}

// Implements GObject::dispose.
static void my_application_dispose(GObject* object) {
  MyApplication* self = MY_APPLICATION(object);
  g_clear_pointer(&self->dart_entrypoint_arguments, g_strfreev);
  G_OBJECT_CLASS(my_application_parent_class)->dispose(object);
}

static void my_application_class_init(MyApplicationClass* klass) {
  G_APPLICATION_CLASS(klass)->activate = my_application_activate;
  G_APPLICATION_CLASS(klass)->local_command_line = my_application_local_command_line;
  G_OBJECT_CLASS(klass)->dispose = my_application_dispose;
}

static void my_application_init(MyApplication* self) {}

MyApplication* my_application_new() {
  return MY_APPLICATION(g_object_new(my_application_get_type(),
                                     "application-id", APPLICATION_ID,
                                     "flags", G_APPLICATION_NON_UNIQUE,
                                     nullptr));
}


example/windows/runner/main.cpp
#include <flutter/dart_project.h>
#include <flutter/flutter_view_controller.h>
#include <windows.h>

#include "flutter_window.h"
#include "utils.h"

int APIENTRY wWinMain(_In_ HINSTANCE instance, _In_opt_ HINSTANCE prev,
                      _In_ wchar_t *command_line, _In_ int show_command) {
  // Attach to console when present (e.g., 'flutter run') or create a
  // new console when running with a debugger.
  if (!::AttachConsole(ATTACH_PARENT_PROCESS) && ::IsDebuggerPresent()) {
    CreateAndAttachConsole();
  }

  // Initialize COM, so that it is available for use in the library and/or
  // plugins.
  ::CoInitializeEx(nullptr, COINIT_APARTMENTTHREADED);

  flutter::DartProject project(L"data");

  std::vector<std::string> command_line_arguments =
      GetCommandLineArguments();

  project.set_dart_entrypoint_arguments(std::move(command_line_arguments));

  FlutterWindow window(project);
  Win32Window::Point origin(10, 10);
  Win32Window::Size size(1280, 720);
  if (!window.CreateAndShow(L"first_packeges", origin, size)) {
    return EXIT_FAILURE;
  }
  window.SetQuitOnClose(true);

  ::MSG msg;
  while (::GetMessage(&msg, nullptr, 0, 0)) {
    ::TranslateMessage(&msg);
    ::DispatchMessage(&msg);
  }

  ::CoUninitialize();
  return EXIT_SUCCESS;
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:ionicons/ionicons.dart';
import 'package:posa_bottom_bar/posa_bottom_bar.dart';
import 'page_test.dart';
// import 'posa_bottom_bar.dart';
// Don't forgot to give me a stars  GITHUB {`HMIDA71`}.

int currentIndex = 0;
void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'POSA BOTTOM BAR EXAMPLE',
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        primarySwatch: Colors.purple,
      ),
      home: const Home(),
    );
  }
}

class Home extends StatefulWidget {
  const Home({Key? key}) : super(key: key);

  @override
  State<Home> createState() => _HomeState();
}

class _HomeState extends State<Home> {
  List pages = [
    const PageTest(),
    const PageTest(),
    const PageTest(),
    const PageTest(),
  ];
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text("POSA BOTTOM BAR"),
        centerTitle: true,
      ),
      body: PosaBottomBar(
        // CHILD IS REQUIRED
        child: pages
            .elementAt(currentIndex), // ADD HERE YOUR PAGE OR PAGES NAVIGATOR
        // ICONS IS REQUIRED
        icons: const [
          Ionicons.home_outline,
          Ionicons.settings_outline,
          Ionicons.bookmark_outline,
          Ionicons.person_outline,
        ],
        // ONCHANGE IS REQUIRED
        onChanged: (i) {
          setState(() {
            debugPrint("TAP $i");
            currentIndex = i;
          });
        },
        // ON LONG PRESS FNC
        onLongPress: (i) {
          setState(() {
            debugPrint("LONG PRESS $i");
          });
        },
        alignment: Alignment.bottomCenter,// Aligment of bottom bar
        // ------ height and margin for bottom bar ----- //
        height: 0.16,
        margin: 0.05,
        // ------  ----- //

        // ------ Elevation & Color for bottom bar ----- //
        elevation: 10,
        shadowColor: Colors.black,
        // ------  ----- //

        borderRadius: 35, // RADUIS (FOR BOTTOM BAR)
        backgroundColor: Colors.blueAccent, // BACKGROUND COLORS (FOR BOTTOM BAR)
        dotColor: Colors.white, // COLORS OF DOT
        iconColor: Colors.white, // COLORS OF ICON
        dotSize: 5.0, iconSize: 24.0, // SIZE OF DOT AND ICONS

        // ------ Animation elevation and opacity /// (`MAKE REFRESH`)  ----- //
        isElevationAnimation: true,
        elevationAnimationOpacity: 0.3,
        elevationAnimationDuration: const Duration(milliseconds: 1500),
        // ------  ----- //

        // ------ Animation Icons ------ //
        isAnimatedIcon: true,
        // ------  ------ //
      ),
    );
  }
}
